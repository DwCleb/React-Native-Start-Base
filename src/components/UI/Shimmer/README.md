```jsx
// TODO: create a Wrapper to react-styleguidist to provide theme to all components shown
import ThemeManager from '../../../theme/ThemeContext';
import Shimmer from '.';

import { View } from 'react-native';

<ThemeManager>
  <Shimmer isLoading={isLoading} style={{ height: 50 }}>
    <View style={{ height: 50 }} />
  </Shimmer>
</ThemeManager>;
```
