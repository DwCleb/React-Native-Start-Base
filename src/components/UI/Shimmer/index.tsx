import React from 'react';
import LinearGradient from 'react-native-linear-gradient';
import { createShimmerPlaceholder } from 'react-native-shimmer-placeholder';
import { useTheme } from '~/theme';
import { ViewStyle } from 'react-native';

export const ShimmerPlaceHolder = createShimmerPlaceholder(LinearGradient);

const defaultShimmerProps = {
  LinearGradient: LinearGradient,
  reverse: false,
  autoRun: true,
  isInteraction: false,
  location: [0.3, 0.5, 0.7],
  duration: 1000,
};

export interface ShimmerProps {
  /* Loading state. */
  isLoading?: boolean;
  /* Wrapped JSX.Elements that will be visible if isLoading = false */
  children?: JSX.Element | JSX.Element[];
  /* Shimmer style */
  style?: ViewStyle;
  /**  A Shimmer is plain when it doesnt wrap a JSX.Element and it's part of
   * a complete page shimmer */
  plain?: boolean;
}

export default function Shimmer({ isLoading, children, style, plain }: ShimmerProps): JSX.Element {
  const { theme } = useTheme();
  return isLoading || plain ? (
    <ShimmerPlaceHolder
      {...defaultShimmerProps}
      visible={plain ? false : !isLoading}
      style={{ ...style, backgroundColor: theme.shimmer.backgroundColor }}
      shimmerColors={[theme.shimmer.primaryColor, theme.shimmer.secondaryColor, theme.shimmer.tertiaryColor]}
    />
  ) : (
    <React.Fragment>{children}</React.Fragment>
  );
}
