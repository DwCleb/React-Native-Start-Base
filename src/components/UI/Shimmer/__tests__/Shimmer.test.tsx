import React from 'react';
import { create, ReactTestRenderer } from 'react-test-renderer';
import Shimmer, { ShimmerProps, ShimmerPlaceHolder } from '../index';
import { ThemeProvider } from 'styled-components/native';
import darkTheme from '~/theme/ThemeContext/dark';
import { Text } from 'react-native';

const theme = darkTheme.theme;

let wrapper: ReactTestRenderer;
let root: ReactTestRenderer['root'];

const childrenElement: JSX.Element = <Text />;

const testProps: ShimmerProps = {
  isLoading: true,
  children: childrenElement,
  style: { width: 100 },
};

const componentToTest = (props: ShimmerProps): JSX.Element => (
  <ThemeProvider theme={theme}>
    <Shimmer {...props} />
  </ThemeProvider>
);

describe('Test for component <Shimmer />', () => {
  beforeAll(() => {
    //ReactNative Animated is not properly mocked so we need to use fakeTimers -> https://github.com/facebook/jest/issues/4359
    jest.useFakeTimers();
  });

  beforeEach(() => {
    wrapper = create(componentToTest(testProps));
    root = wrapper.root;
  });

  it('Should renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('Should render correctly its props', () => {
    const ShimmerComponent = root.findAllByType(ShimmerPlaceHolder)[0];
    expect(ShimmerComponent).toBeTruthy();
    expect(ShimmerComponent.props.visible).toEqual(!testProps.isLoading);
    expect(ShimmerComponent.props.style).toEqual({
      backgroundColor: theme.shimmer.backgroundColor,
      ...testProps.style,
    });
  });

  it('Should render its children JSX.Elements when loading=false', () => {
    wrapper = create(componentToTest({ ...testProps, isLoading: false }));
    root = wrapper.root;

    const ChildrenComponent = root.findAllByType(Text);
    const ShimmerComponent = root.findAllByType(ShimmerPlaceHolder);

    expect(ShimmerComponent).toHaveLength(0);
    expect(ChildrenComponent).toBeTruthy();
    expect(ChildrenComponent).toHaveLength(1);
  });

  it('Shimmer should be visible (visible=false) when plain flag is truthy', () => {
    wrapper = create(componentToTest({ plain: true, children: <Text /> }));
    root = wrapper.root;

    const ShimmerComponent = root.findAllByType(ShimmerPlaceHolder)[0];
    const ChildrenComponent = root.findAllByType(Text);

    expect(ShimmerComponent.props.visible).toBeFalsy();
    expect(ChildrenComponent).toHaveLength(0);
  });
});
