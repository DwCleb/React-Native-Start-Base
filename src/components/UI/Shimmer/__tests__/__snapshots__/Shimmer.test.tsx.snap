// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test for component <Shimmer /> Should renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "height": 15,
        "width": 200,
      },
      Object {
        "overflow": "hidden",
      },
      undefined,
      Object {
        "backgroundColor": "#3A3D43",
        "width": 100,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "height": 0,
          "opacity": 0,
          "width": 0,
        },
        false,
      ]
    }
  />
  <View
    style={
      Object {
        "backgroundColor": "#3D4046",
        "flex": 1,
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
          "transform": Array [
            Object {
              "translateX": -200,
            },
          ],
        }
      }
    >
      <BVLinearGradient
        colors={
          Array [
            4282204230,
            4282006851,
            4282204230,
          ]
        }
        endPoint={
          Object {
            "x": 2,
            "y": 0.5,
          }
        }
        locations={
          Array [
            0.3,
            0.5,
            0.7,
          ]
        }
        startPoint={
          Object {
            "x": -1,
            "y": 0.5,
          }
        }
        style={
          Object {
            "flex": 1,
            "width": 200,
          }
        }
      />
    </View>
  </View>
</View>
`;
